rust_src/crates/lisp/src/eval.rs

```diff
@@ -26,3 +36,94 @@ pub fn signal_rust(error_symbol: LispObject, data: LispObject) -> ! {
         unreachable!();
     }
 }
+
+enum LispFun {
+    SubrFun(LispSubrRef),
+    LambdaFun(LispObject),
+}
+
+enum LispFunError {
+    InvalidFun,
+    VoidFun,
+}
+
+/// Call first argument as a function, passing remaining arguments to it.
+/// Return the value that function returns.
+/// Thus, (funcall \\='cons \\='x \\='y) returns (x . y).
+/// usage: (funcall FUNCTION &rest ARGUMENTS)
+#[allow(unused_assignments)]
+pub fn funcall(args: &mut [LispObject]) -> LispObject {
```

Comment:
Here's my take on this - I think we will get ourselves into trouble long term if we copy too much of remacs re-implementations. I think copying remacs wrapper classes, and simple one liners (like intern below) are good, but larger functions like funcall will cause issues. Instead, I think it would be better if we made these functions 'safe wrappers' around the native C functions - that way we keep the implementations aligned, but we are still writing idiomatic Rust and aren't littering the code with unsafe and `as_mut_ptr()` calls .

Reply:
Yeah I was also hesitant to do this. I'll look for an alternative.